feat: update cognito login guide for realworld app (#4882)
cypress-io/cypress-documentation: content/guides/end-to-end-testing/amazon-cognito-authentication.md

## <Icon name="graduation-cap"></Icon> What you'll learn

> - Authenticate with [`cy.origin()`](/api/commands/origin) or programmatically
>   (legacy) to [Amazon Cognito](https://aws.amazon.com/cognito) via a custom

Reviewer: ```suggestion
- Login to [Amazon Cognito](https://aws.amazon.com/cognito) through the UI with [`cy.origin()`](/api/commands/origin)
- Programmatically authenticate with [Amazon Cognito](https://aws.amazon.com/cognito) via a custom
```

Reviewer: Not sure if we had a stance, but programmatic login is a valid flow and I don't necessarily fee like we should label it legacy. I think it would just depend on what makes the most sense for a user's test suite and CI setup.

Author: I think thats fair. I feel like we would want to encourage the use of `origin` over programmatic due to how involved programmatic can be, but it is for sure a valid flow and probably something we don't have to call legacy

Author: addressed in e14bf7170

  Cypress command

> - If needed, adapting your [Amazon Cognito](https://aws.amazon.com/cognito)
>   application for programmatic authentication during testing

Reviewer: ```suggestion
- Adapting your [Amazon Cognito](https://aws.amazon.com/cognito)
  application for programmatic authentication during testing
```

Author: addressed in e14bf7170

</Alert>

<Alert type="success">

<strong class="alert-header">Authenticate by visiting a different domain with
[`cy.origin()`](/api/commands/origin)</strong>

> Programmatic authentication for Cognito is now considered a legacy

Reviewer: I think it's good to explain why we are recommending origin first
```suggestion
Typically, logging in a user within your app by authenticating via a third-party provider requires visiting login pages hosted on a different domain. Before Cypress 12, Cypress tests were limited to visiting domains of the same origin, so programatic login was the only option for authenticating with a third-party authentication API. 
```

Author: Took a stab at this in e14bf7170

## <Icon name="graduation-cap"></Icon> What you'll learn

> - Login to [Amazon Cognito](https://aws.amazon.com/cognito) through the UI with

Reviewer 2: ```suggestion
- Log in to [Amazon Cognito](https://aws.amazon.com/cognito) through the UI with
```

Author: addressed in 8bec0b5d1

  [`cy.origin()`](/api/commands/origin)
- Programmatically authenticate with

>   [Amazon Cognito](https://aws.amazon.com/cognito) via a custom command Cypress
>   command

Author: ```suggestion
  [Amazon Cognito](https://aws.amazon.com/cognito) via a custom Cypress
  command
```

## Custom Command for Amazon Cognito Authentication

There are two ways you can authenticate to AWS Cognito

- [Login with cy.origin()](/guides/end-to-end-testing/amazon-cognito-authentication#Login-with-cy-origin)
- [Legacy Programmatic Access](/guides/end-to-end-testing/amazon-cognito-authentication#Programmatic-Login-Legacy)

### Login with [`cy.origin()`](/api/commands/origin)

> Next, we'll write a command to perform a login to

Reviewer: ```suggestion
Next, we'll write a custom command called, `loginByCognito`, to perform a login to
```

[Amazon Cognito](https://aws.amazon.com/cognito) using

> [`cy.origin()`](/api/commands/origin) to navigate to the cognito origin,
> inputting a user credentials, and signing in via OAuth to redirect back to our

Reviewer: ```suggestion
[`cy.origin()`](/api/commands/origin) to:
1.  navigate to the cognito origin
2. inputting a user credentials
3. signing in via OAuth which will redirect back to our application
4. cache the results with
[`cy.session()`](/api/commands/session).
```
just an idea. I know when I'm learning, I only skim content instead of read in depth

Author: addressed in e14bf7170

application, and caching the results with
[`cy.session()`](/api/commands/session).

> In this `loginByCognito` command, we redirect to the cognito login screen by

Reviewer: This seems to be repeating the above snipped with a little more details

Author: removed in e14bf7170

authenticating with `Sign in with AWS`. We then enter our user credentials and
sign in, which redirects us back to the realworld app.

```jsx
// cypress/support/auth-provider-commands/cognito.ts

/// Amazon Cognito
Cypress.Commands.add('loginByCognito', (username, password) => {
  Cypress.log({
    displayName: 'COGNITO LOGIN',
    message: [`🔐 Authenticating | ${username}`],

>     // @ts-ignore

Reviewer: ```suggestion
```

Author: removed in e14bf7170

    autoEnd: false,
  })

  cy.visit('/')
  cy.contains('Sign in with AWS', {

>     includeShadowDom: true,

Reviewer: just curious, is this needed?

Author: the cognito react components are all shadow DOM roots 🥴

  }).click()

  cy.origin(
    Cypress.env('cognito_domain'),
    {
      args: {
        username,
        password,
      },
    },
    ({ username, password }) => {
      // cognito log in page has some elements of the same id but are off screen. we only want the visible elements to log in
      cy.get('input[name="username"]:visible').type(username)
      cy.get('input[name="password"]:visible').type(password, {

>         log: false,

Reviewer: ```suggestion
        log: false, // use log: false to prevent your password from showing in the Command Log
```

Author: added in e14bf7170

      })
      cy.get('input[name="signInSubmitButton"]:visible').click()
    }
  )

  // give a few seconds for redirect to settle

>   // eslint-disable-next-line cypress/no-unnecessary-waiting

Reviewer: ```suggestion
```

Author: removed in e14bf7170

  cy.wait(2000)

  // verify we have made it passed the login screen
  cy.contains('Get Started').should('be.visible')
})
```

Secondly, we can use our `loginByCognito` command in the test. Below is our test
to login as a user via [Amazon Cognito](https://aws.amazon.com/cognito),
complete the onboarding process and logout.

<Alert type="success">

The
[runnable version of this test](https://github.com/cypress-io/cypress-realworld-app/blob/develop/cypress/tests/ui-auth-providers/cognito.spec.ts)
is in the
[Cypress Real World App](https://github.com/cypress-io/cypress-realworld-app).

</Alert>

```jsx
describe('Cognito', function () {
  beforeEach(function () {
    // Seed database with test data
    cy.task('db:seed')

    // login via Amazon Cognito via cy.origin()
    cy.loginByCognito(
      Cypress.env('cognito_username'),
      Cypress.env('cognito_password')

>     )

Reviewer: ever use the diff markdown language? not sure what we typically do, but could highlight like:
```diff
  beforeEach(function () {
    // Seed database with test data
    cy.task('db:seed')
    // login via Amazon Cognito via cy.origin()
+    cy.loginByCognito(
+      Cypress.env('cognito_username'),
+      Cypress.env('cognito_password')
+    )
```

Author: I haven't. That's neat!

Reviewer: really? its my fav 😄

Author: It does a look a bit weird with the docs markdown renderer. Doesn't look like we get the cool green highlights:
<img width="849" alt="Screen Shot 2022-12-01 at 11 13 22 AM" src="https://user-images.githubusercontent.com/3980464/205103544-049bf1b2-b375-4e39-a165-658e588c4cb6.png">

Reviewer: @AtofStryker did you use ` ```js ` or ` ```diff`? looks like it's using JS/jsx

Reviewer: it's either JS or diff, not both :/ so JSX might be clearer.

Author: might just stick with the jsx on this one. the contract just looks off 
<img width="690" alt="Screen Shot 2022-12-01 at 12 50 19 PM" src="https://user-images.githubusercontent.com/3980464/205124717-deb9d860-5334-49b6-a630-b25b9ddb6f44.png">

  })

  it('shows onboarding', function () {
    cy.contains('Get Started').should('be.visible')
  })
})
```

<DocsVideo src="/img/examples/aws-cognito-origin.mp4"></DocsVideo>

Lastly, we can refactor our login command to take advantage of
[`cy.session()`](/api/commands/session) to store our logged in user so we don't
have to reauthenticate with everything test.

```jsx
// Amazon Cognito
Cypress.Commands.add('loginByCognito', (username, password) => {

>   cy.session(

Reviewer: This example is pretty long. I am familiar with sessions & origin to see the changes, but it might be hard for newbies. Wondering if we could set it up like:
```
// first example
Cypress.Commands.add('loginByCognito', (username, password) => {
    const login = () => {
      cy.origin()...
    } 
    return login()
}
// second example

Cypress.Commands.add('loginByCognito', (username, password) => {
    const login = () => {
      cy.origin()...
    } 
   cy.session(`cognito-${username}`, login)
}
```

Author: a lot easier to parse this way and we don't nearly need to repeat as much

Author: addressed in e14bf7170

      cy.contains('Sign in with AWS', {
        includeShadowDom: true,
      }).click()

      cy.origin(
        Cypress.env('cognito_domain'),
        {
          args: {
            username,
            password,
          },
        },
        ({ username, password }) => {
          // cognito log in page has some elements of the same id but are off screen. we only want the visible elements to log in
          cy.get('input[name="username"]:visible').type(username)
          cy.get('input[name="password"]:visible').type(password, {
            log: false,
          })
          cy.get('input[name="signInSubmitButton"]:visible').click()
        }
      )

      // give a few seconds for redirect to settle
      // eslint-disable-next-line cypress/no-unnecessary-waiting
      cy.wait(2000)

      // verify we have made it passed the login screen
      cy.contains('Get Started').should('be.visible')
    },
    {
      validate() {
        cy.visit('/')
        // revalidate our session to make sure we are logged in
        cy.contains('Get Started').should('be.visible')
      },
    }
  )
})
```

<DocsVideo src="/img/examples/cognito-session-restore.mp4"></DocsVideo>

> Unlike programmatic login, authenticating with

Reviewer: ah maybe jumping the gun? I haven't ready about programatic login yet! Maybe move this to the section `Adapting an Amazon Cognito App for Testing`.

We could all out there, if you want to use cy.orgion, you can skip this section. and/or maybe we can update the title to say something like `Adapting an Amazon Cognito App for Programmatic Login`.

Author: I added a little info disclaimer in the section instead of changing the title in e14bf7170. How does this look?

<img width="690" alt="Screen Shot 2022-12-01 at 12 44 49 PM" src="https://user-images.githubusercontent.com/3980464/205123659-f89187bd-0063-41c2-9e9c-2eb32d01e32d.png">

[`cy.origin()`](/api/commands/origin) does not require adapting the application
to work. It simply works without intervention exactly how your users would
consume the application!

> ### Programmatic Login (Legacy)

Reviewer: ```suggestion
### Programmatic Login
```

Author: addressed in e14bf7170

})
```

> #### Adapting an Amazon Cognito App for Testing

Reviewer: ```suggestion
## Adapting an Amazon Cognito App for Testing
```

Reviewer: Is this required for both logins or only for using programatic login?  If the later, we should specify

Author: only for programmatic, which is why I changed the header values to fall under the programmatic section. We can keep those the same, but it does feel a bit strange to call that out in each subsection

Reviewer: Fair, we had called it out above which is why I had originally thought it should be it's own section

Author: I updated the header in e14bf7170 so the navbar shows the adapting section as such:
<img width="262" alt="Screen Shot 2022-12-01 at 12 41 00 PM" src="https://user-images.githubusercontent.com/3980464/205123457-7bc5eb2a-a11f-4c3b-9446-7427a46c49b2.png">
